File: api_server.py
import datetime

from flask import Flask, jsonify, request
from influxdb_client import InfluxDBClient

from config import *

app = Flask(__name__)


# Initialize InfluxDB client
client = InfluxDBClient(url=INFLUXDB_URL, token=INFLUXDB_TOKEN)
query_api = client.query_api()


def query_influxdb(query):
    """Run a query against InfluxDB and return the results."""
    try:
        result = query_api.query(org=INFLUXDB_ORG, query=query)
        return result
    except Exception as e:
        print(f"InfluxDB query failed: {e}")
        return None


def _get_period_from_request() -> (int, int, dict):
    error = {}
    start_minutes = end_minutes = 0

    try:
        start_minutes = int(request.args.get("start", 1))
    except ValueError:
        error = {"start": "start parameter must be an integer"}

    try:
        end_minutes = int(request.args.get("end", 0))
    except ValueError:
        error = {"end": "start parameter must be an integer"}

    if start_minutes < 1:
        # There should be proper error messages if this was a real API.
        start_minutes = 1

    if end_minutes < 0:
        # There should be proper error messages if this was a real API.
        end_minutes = 0

    if start_minutes < end_minutes:
        error = {"start": "start value should be larger than end value"}

    return start_minutes, end_minutes, error


@app.route("/metrics", methods=["GET"])
def get_metrics():
    """Retrieve metrics for a specific time range defined by an integer in minutes, going back from now.
    Use as:
    curl http://localhost:5000/metrics?start=10&end=2
    To get metrics starting 10 minutes back going to 2 minutes back; end is optional, defaults to zero (now).

    InfluxDB does not seem to support proper "from" and "to" values with actual times.
    @see https://community.influxdata.com/t/query-specific-time-range-and-day/14503/8
    """
    measurement = request.args.get("measurement", "system_metrics")

    start_minutes, end_minutes, error = _get_period_from_request()
    if error:
        return jsonify(error), 400

    query = f"""
        from(bucket: "{INFLUXDB_BUCKET}")
            |> range(start: -{start_minutes}m, stop: -{end_minutes}m)
            |> filter(fn: (r) => r["_measurement"] == "{measurement}")
        """

    if VERBOSE:
        print(f"Executing query: {query}")

    try:
        result = query_influxdb(query)
    except Exception as e:
        return jsonify({"error": f"Query execution failed: {e}"}), 500

    if VERBOSE:
        print(f"Query result: {result}")

    if not result:
        return jsonify({"error": "No results found"}), 404

    data = []
    for table in result:
        for record in table.records:
            data.append(
                {
                    "measurement": record.get_measurement(),
                    "time": record.get_time(),
                    "field": record.get_field(),
                    "value": record.get_value(),
                }
            )
    return jsonify(data)


@app.route("/metrics/aggregate", methods=["GET"])
def aggregate_metrics():
    """Aggregate metrics, e.g., average memory usage over the last hour.
    Handling CPU properly will require more parameters, because on multicore it splits into e.g. eight sub-values, and
    handling that is quite out of scope, so we have a crude list below which may blow up if it runs on 4 CPUs and not 8.

    Use as:
    curl http://localhost:5000/metrics/aggregate?metric=memory_usage&start=10&end=2
    To get metrics starting 10 minutes back going to 2 minutes back; end is optional, defaults to zero (now).

    This only gets the mean value now, expanding it with other aggregations is an exercise for the reader.
    """
    measurement = request.args.get("measurement", "system_metrics")
    metric = request.args.get("metric")
    # These may belong in the config file.
    allowed_metrics = [
        "cpu_usage_overall",
        "cpu_usage_1",
        "cpu_usage_2",
        "cpu_usage_3",
        "cpu_usage_4",
        "cpu_usage_5",
        "cpu_usage_6",
        "cpu_usage_7",
        "cpu_usage_8",
        "memory_usage",
        "disk_read",
        "disk_write",
        "network_sent",
        "network_recv",
    ]
    if metric not in allowed_metrics:
        return (
            jsonify({"metric": f"Metric not in allowed list: {allowed_metrics}"}),
            400,
        )

    start_minutes, end_minutes, error = _get_period_from_request()
    if error:
        return jsonify(error), 400

    query = f"""
    from(bucket: "{INFLUXDB_BUCKET}")
        |> range(start: -1h)
        |> filter(fn: (r) => r["_measurement"] == "{measurement}")
        |> filter(fn: (r) => r["_field"] == "{metric}")
        |> mean()
    """
    result = query_influxdb(query)

    if not result:
        return jsonify({"error": "Failed to query InfluxDB"}), 500

    data = []
    # With the current logic, there will always be just one result.
    for table in result:
        for record in table.records:
            data.append(
                {
                    "measurement": record.get_measurement(),
                    "field": record.get_field(),
                    "value": record.get_value(),
                }
            )
    return jsonify(data)


if __name__ == "__main__":
    app.run(debug=True)

File: api_server.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"github.com/influxdata/influxdb-client-go/v2"
    "github.com/influxdata/influxdb-client-go/v2/api"
)

var (
	INFLUXDB_URL    = "http://localhost:8086"
	INFLUXDB_TOKEN  = "4oYGSdMFbjiZU3JljF5_UxXxnsvIxTN2bHP24gfcTFCtfymXUxXmvDaraGbl9K5OqfVTpTaDNwCGZwAg4XAKBQ=="
	INFLUXDB_ORG    = "pavka"
	INFLUXDB_BUCKET = "rhtest"
)

var client = influxdb2.NewClient(INFLUXDB_URL, INFLUXDB_TOKEN)
var queryAPI = client.QueryAPI(INFLUXDB_ORG)

func getPeriodFromRequest(r *http.Request) (start int, end int, err error) {
	startStr := r.URL.Query().Get("start")
	endStr := r.URL.Query().Get("end")

	if startStr == "" {
		start = 1
	} else {
		start, err = strconv.Atoi(startStr)
		if err != nil {
			return 0, 0, fmt.Errorf("invalid start value: %v", err)
		}
	}

	if endStr == "" {
		end = 0
	} else {
		end, err = strconv.Atoi(endStr)
		if err != nil {
			return 0, 0, fmt.Errorf("invalid end value: %v", err)
		}
	}

	if start < end {
		err = fmt.Errorf("start value should be larger than end value")
		return 0, 0, err
	}

	return
}

func queryInfluxDB(query string) (*api.QueryTableResult, error) {
	result, err := queryAPI.Query(context.Background(), query)
	if err != nil {
		return nil, err
	}
	return result, nil
}

func getMetricsHandler(w http.ResponseWriter, r *http.Request) {
    measurement := r.URL.Query().Get("measurement")
    if measurement == "" {
        measurement = "system_metrics"
    }

    start, end, err := getPeriodFromRequest(r)
    if err != nil {
        http.Error(w, err.Error(), http.StatusBadRequest)
        return
    }

    query := fmt.Sprintf(`
        from(bucket: "%s")
            |> range(start: -%dm, stop: -%dm)
            |> filter(fn: (r) => r["_measurement"] == "%s")
    `, INFLUXDB_BUCKET, start, end, measurement)

    result, err := queryInfluxDB(query)
    if err != nil {
        http.Error(w, fmt.Sprintf("Failed to query InfluxDB: %v", err), http.StatusInternalServerError)
        return
    }

    // Convert the result into a JSON-serializable structure
    data := make([]map[string]interface{}, 0)
    for result.Next() {
        record := result.Record()
        row := map[string]interface{}{
            "time":        record.Time(),
            "measurement": record.Measurement(),
            "field":       record.Field(),
            "value":       record.Value(),
        }
        data = append(data, row)
    }

    if result.Err() != nil {
        http.Error(w, fmt.Sprintf("Error processing result: %v", result.Err()), http.StatusInternalServerError)
        return
    }

    // Set the response content type to application/json
    w.Header().Set("Content-Type", "application/json")
    w.WriteHeader(http.StatusOK)

    // Serialize the data to JSON and write it to the response
    if err := json.NewEncoder(w).Encode(data); err != nil {
        http.Error(w, fmt.Sprintf("Failed to serialize data to JSON: %v", err), http.StatusInternalServerError)
        return
    }
}

func getMetricsAggregateHandler(w http.ResponseWriter, r *http.Request) {
    measurement := r.URL.Query().Get("measurement")
    if measurement == "" {
        measurement = "system_metrics"
    }

    metric := r.URL.Query().Get("metric")
    if metric == "" {
        http.Error(w, "Missing 'metric' parameter", http.StatusBadRequest)
        return
    }

    start, end, err := getPeriodFromRequest(r)
    if err != nil {
        http.Error(w, err.Error(), http.StatusBadRequest)
        return
    }

    query := fmt.Sprintf(`
        from(bucket: "%s")
            |> range(start: -%dm, stop: -%dm)
            |> filter(fn: (r) => r["_measurement"] == "%s")
            |> filter(fn: (r) => r["_field"] == "%s")
            |> mean()
    `, INFLUXDB_BUCKET, start, end, measurement, metric)

    result, err := queryInfluxDB(query)
    if err != nil {
        http.Error(w, fmt.Sprintf("Failed to query InfluxDB: %v", err), http.StatusInternalServerError)
        return
    }

    var value interface{}
    if result.Next() {
        value = result.Record().Value()
    } else {
        http.Error(w, "No data found", http.StatusNotFound)
        return
    }

    if result.Err() != nil {
        http.Error(w, fmt.Sprintf("Error processing result: %v", result.Err()), http.StatusInternalServerError)
        return
    }

    // Set the response content type to application/json
    w.Header().Set("Content-Type", "application/json")
    w.WriteHeader(http.StatusOK)

    // Serialize the value to JSON and write it to the response
    response := map[string]interface{}{
        "value": value,
    }

    if err := json.NewEncoder(w).Encode(response); err != nil {
        http.Error(w, fmt.Sprintf("Failed to serialize data to JSON: %v", err), http.StatusInternalServerError)
        return
    }
}

func main() {
	http.HandleFunc("/metrics", getMetricsHandler)
	http.HandleFunc("/metrics/aggregate", getMetricsAggregateHandler)

	http.ListenAndServe(":5000", nil)
}
